// irc/proto.js
/*
Describe messages sent by the server and how to read them
:sender command params\r\n
command is the key, params described in the value (bounds are implied)
Everything's literally matched except:
{name} matches a value potentially with spaces and saves to name (.*)
{name:N} same as above, but may only contain up to N spaces
{name=N} same as above, but must contain exactly N spaces
<name> matches a value without spaces and saves to name ([^ ]*)
<name:etc> same as above, but characters in etc are also not allowed
+Note that to use a literal - instead of a range, use --
<name=etc> match a value that only contains etc. Put a + on the end
+to match multiple of those characters.
<name!etc> same as above but negative.
<#name> matches number and saves to name ([0-9]+)
<#name:.,> matches number with given modifiers ([0-9,]+(\.[0-9]+)?)
+Put an s on the end to switch the meaning of . and ,
<name@otherName> inherets the match rules of otherName. Preceding
+otherName with a _ references the global rule, otherwise references
+one in the current message. Naming is nested for globals.
/name:regex/ specific match (grouping is implied) note: regex is JS's
suffixing with a * (ie. directly after } > or /) will cause the same
+thing to match multiple times and return an array under that name
escape {}<>/%* by preceeding it with a %
Can be an array for multiple forms.
Sections with the same name will assert equivalence within that key
You may denote a section with ()
Sections surrounded by [] denote optional sections.
If there are multiple options per section use [grp1]|[grp2]
If they're not optional, use (grp1)|(grp2)
For all groups - {} <> // () [] - you may start it with one character
+followed by a | to denote what the separator is between multiple
+instances of this section. By default it's space
*/

/*
Message Object {
	sc: SERVER | CLIENT, //sent from Server and/or Client
	//Can be arrays
	msg: "", //Message passed to splitter, default
	rsp: [], //Possible response commands, default
	//The below can also be strings, assumed to be msg
	rfc1459 { //only using rfc1459 (this before default)
		msg: "",
		rsp: [],
	},
	rfc281x: { //with rfc2810-rfc2813 enabled
		msg: "",
		rsp: [],
	},
}
*/

var Splitter = require("splitter.js");

var SERVER=1,CLIENT=2,INTERSERVER=4,msgs;

function splitWholeEntries(len,data){
	var ret=[];
	while(data.length){
		var rlen = -1;
		for(var i=0;i<data.length && (rlen+=(data[i].length+1)) <= len;++i);
		if(i == 0)
		ret.push(data.slice(0,i));
		data = data.slice(i);
	}
	return ret;
}

function splitFormattedText(len,data){
	var ret = [];
	while(data.length){
		//Make sure to carry over formatting and CTCPs if they can't be completed on that line
		//Word wrap, as well
		//Reference: CTCP \1, Bold \2, Color \3, Italic/Reversed \26, Underline \37, Unformat \17
		var rstr = data.substr(0,len)
		,   rmdr = data.substr(len)
		,   pos,fmt;
		if(rstr.match(/\1/g).length % 2 == 1){
			//CTCP was cut off
			pos = rstr.lastIndexOf("\1");
			rmdr = rstr.substr(pos)+rmdr;
			rstr = rstr.substr(0,pos);
		}
		//Split to last bound (if it's not too far)
		var tmp;
		if(tmp=rstr.match(/(.*[^a-zA-Z0-9\1\2\3\26\37\17].*?)([^\1]*)$/)
		&& tmp[2].length < 20 && !rmdr.match(/^[^a-zA-Z0-9\1\2\3\26\37\17]/)){
			rstr = tmp[1];
			rmdr = tmp[2]+rmdr;
		}
		fmt = rstr.match(/[\2\3\26\37\17]/g);
		if((pos = fmt.lastIndexOf("\17")) != -1){
			//Stopper was found, only retain what comes after
			fmt.slice(pos+1);
		}
		if(fmt.length){
			//Formatting left over, prepend to remainder
			rmdr = fmt.join("")+rmdr;
		}

		ret.push(rstr)
		data = rmdr;
	}
	return ret;
}

module.exports = msgs = {
	"_server": ":<sender> <command> {args}",
	"_client": "<command> {args}",
	"_numeric": "<nick> {args}",
	//"_args": "<spaced>* :{trailing}",
	"_target": "(,|"+
		"(/prefix:[#+&]|![A-Z0-9]{5}/<channel:\0\7\r\n,:>[:<mask@channel>])|"+
		"(<user>[%<host>]@<server>)|"+
		"(<user>%<host>)|"+
		"(<nick>[!<user>@<host>])|"+
		//"([$#]"
	")*",
	"_hostname": "<,|@_shortname>*",
	"_shortname": "/:[a-zA-Z0-9][a-zA-Z0-9--][a-zA-Z0-9]/",
	"001": {
		sc: SERVER,
		msg: ":Welcome to the {network} <nick>!<user>@<host>",
	},
	"002": {
		sc: SERVER,
		msg: ":Your host is <serverName>, running version <version>",
	},
	"003": {
		sc: SERVER,
		msg: ":This server was created {date}",
	},
	"004": {
		sc: SERVER,
		msg: "<serverName> <version> <okUserModes> <okChannelModes>",
	},
	"005": [{
		sc: SERVER,
		msg: "<support>* :are supported by this server", //Sent at the start
		split: ["support",splitWholeEntries],
	},{
		sc: SERVER,
		msg: ":Try server <server>, port <#port>", //RPL_BOUNCE (RFC2812)
	}],
	
	//NOTE: All of these with msg only are from rfc1459, check rfc281x for differences.
	"200": { //NOTE: Correct?
		sc: SERVER,
		msg: "Link <version>[.<debugLevel>] <destination> <next server>",
	},
	"201": {
		sc: SERVER,
		msg: "Try. <class> <server>",
	},
	"202": {
		sc: SERVER,
		msg: "H.S. <class> <server>",
	},
	"203": {
		sc: SERVER,
		msg: "???? <class>[ <clientIP>]",
	},
	"204": {
		sc: SERVER,
		msg: "Oper <class> <nick>",
	},
	"205": {
		sc: SERVER,
		msg: "User <class> <nick>",
	},
	"206": {
		sc: SERVER,
		msg: "Serv <class> <#S>S <#C>C <server> <nick>!<user>@<host>",
	},
	"208": {
		sc: SERVER,
		msg: "<newtype> 0 <clientName>",
	},
	"211": {
		sc: SERVER,
		msg: "<linkname> <sendq> <sentMessages> <sentBytes> <receivedMessages> <receivedBytes> <timeOpen>",
	},
	"212": {
		sc: SERVER,
		msg: "<command> <count>",
	},
	"213": {
		sc: SERVER,
		msg: "C <host> * <name> <port> <class>",
	},
	"214": {
		sc: SERVER,
		msg: "N <host> * <name> <port> <class>",
	},
	"215": {
		sc: SERVER,
		msg: "I <host> * <host> <port> <class>",
	},
	"216": {
		sc: SERVER,
		msg: "K <host> * <username> <port> <class>",
	},
	"218": {
		sc: SERVER,
		msg: "Y <class> <pingFrequency> <connectFrequency> <maxSendq>",
	},
	"219": {
		sc: SERVER,
		msg: "<letter> :End of /STATS report",
	},
	"241": {
		sc: SERVER,
		msg: "L <hostmask> * <servername> <maxdepth>",
	},
	"242": {
		sc: SERVER,
		msg: ":Server Up <#days> days <#hour>:<#min>:<#sec>",
	},
	"243": {
		sc: SERVER,
		msg: "O <hostmask> * <name>",
	},
	"244": {
		sc: SERVER,
		msg: "H <hostmask> * <servername>",
	},
	"221": {
		sc: SERVER,
		msg: "<userModes>",
	},
	"251": {
		sc: SERVER,
		msg: ":There are <#userCount> users and <#invisCount> invisible on <#servCount> servers",
	},
	"252": {
		sc: SERVER,
		msg: "<#opsCount> :operator%(s%) online",
	},
	"253": {
		sc: SERVER,
		msg: "<#unknownConns> :unknown connection%(s%)",
	},
	"254": {
		sc: SERVER,
		msg: "<#chanCount> :channels formed",
	},
	"255": {
		sc: SERVER,
		msg: ":I have <#clients> clients and <#servers> servers",
	},
	"256": [{
		sc: SERVER,
		msg: "<server> :Administrative info",
	},{
		sc: SERVER,
		msg: "[<server> ]:Administrative info about <server>",
	}]
	"257": {
		sc: SERVER,
		msg: ":{adminLocation1}?",
	},
	"258": {
		sc: SERVER,
		msg: ":{adminLocation2}?",
	},
	"259": {
		sc: SERVER,
		msg: ":{adminEmail}?",
	},
	"261": {
		sc: SERVER,
		msg: "File <logfile> <debug level>",
	},
	
	"300": {
		sc: SERVER,
		msg: "[{data}]",
	},
	"302": {
		sc: SERVER,
		msg: ":[<nick:*=>/op:%*?/=<away=+-><host>]*",
	},
	"301": {
		sc: SERVER,
		msg: "<nick> :{awayMessage}",
	},
	"303": {
		sc: SERVER,
		msg: ":<nick>*?",
	},
	"305": {
		sc: SERVER,
		msg: ":You are no longer marked as being away",
	},
	"306": {
		sc: SERVER,
		msg: ":You have been marked as being away",
	},
	"311": {
		sc: SERVER,
		msg: "<nick> <user> <host> * :{realName}",
	},
	"312": {
		sc: SERVER,
		msg: "<nick> <server> :{serverInfo}",
	},
	"313": {
		sc: SERVER,
		msg: "<nick> :is an IRC operator",
	},
	"314": {
		sc: SERVER,
		msg: "<nick> <user> <host> * :{realName}",
	},
	"315": {
		sc: SERVER,
		msg: "<name> :End of /WHO list",
	},
	"317": {
		sc: SERVER,
		msg: "<nick> <#seconds> :seconds idle",
	},
	"318": {
		sc: SERVER,
		msg: "<nick> :End of /WHOIS list",
	},
	"319": {
		sc: SERVER,
		msg: "<nick> :<channel>*",
	},
	"321": {
		sc: SERVER,
		msg: "Channel :Users<= +>Name",
	},
	"322": {
		sc: SERVER,
		msg: "<channel> <#userCount> :{topic}",
	},
	"323": {
		sc: SERVER,
		msg: ":End of /LIST",
	},
	"324": {
		sc: SERVER,
		msg: "<channel> <modes> <modeParams>*",
	},
	"331": {
		sc: SERVER,
		msg: "<channel> :No topic is set",
	},
	"332": {
		sc: SERVER,
		msg: "<channel> :<topic>",
	},
	"341": {
		sc: SERVER,
		msg: "<nick> <channel>",
	},
	"342": {
		sc: SERVER,
		msg: "<user> :Summoning user to IRC",
	},
	"351": {
		sc: SERVER,
		msg: "<version>[.<debuglevel>] <server> :{comments}",
	},
	"352": {
		//TODO: What does the asterisk mean? I see an r sometimes, is that the status? What is HG?
		sc: SERVER,
		msg: "<channel> <user> <host> <server> <nick> <HG=HG>/asterisk:*//status:.?/ :<hopcount> {realName}",
	},
	"353": {
		sc: SERVER,
		msg: "<channel> :<nicks>*",
	},
	//TODO: fix below's syntax
	"364": {
		sc: SERVER,
		msg: "<mask> <server> :<hopcount> <server info>",
	},
	"365": {
		sc: SERVER,
		msg: "<mask> :End of /LINKS list",
	},
	"366": {
		sc: SERVER,
		msg: "<channel> :End of /NAMES list",
	},
	"367": {
		sc: SERVER,
		msg: "<channel> <banid>",
	},
	"368": {
		sc: SERVER,
		msg: "<channel> :End of channel ban list",
	},
	"369": {
		sc: SERVER,
		msg: "<nick> :End of WHOWAS",
	},
	"371": {
		sc: SERVER,
		msg: ":<string>",
	},
	"372": {
		sc: SERVER,
		msg: ":- <text>",
	},
	"374": {
		sc: SERVER,
		msg: ":End of /INFO list",
	},
	"375": {
		sc: SERVER,
		msg: ":- <server> Message of the day - ",
	},
	"376": {
		sc: SERVER,
		msg: ":End of /MOTD command",
	},
	"381": {
		sc: SERVER,
		msg: ":You are now an IRC operator",
	},
	"382": {
		sc: SERVER,
		msg: "<config file> :Rehashing",
	},
	"391": {
		sc: SERVER,
		msg: "<server> :<string showing server's local time>",
	},
	"392": {
		sc: SERVER,
		msg: ":UserID Terminal Host",
	},
	"393": {
		sc: SERVER,
		msg: ":%-8s %-9s %-8s",
	},
	"394": {
		sc: SERVER,
		msg: ":End of users",
	},
	"395": {
		sc: SERVER,
		msg: ":Nobody logged in",
	},
	//TODO: fix above's syntax
	"376": {
		sc: SERVER,
		msg: ":End of /MOTD command",
	},
	"401": {
		sc: SERVER,
		msg: "<nick> :No such nick/channel",
	},
	"402": {
		sc: SERVER,
		msg: "<server> :No such server",
	},
	"403": {
		sc: SERVER,
		msg: "<channel> :No such channel",
	},
	"404": {
		sc: SERVER,
		msg: "<channel> :Cannot send to channel",
	},
	"405": {
		sc: SERVER,
		msg: "<channel> :You have joined too many channels",
	},
	"406": {
		sc: SERVER,
		msg: "<nick> :There was no such nickname",
	},
	"407": {
		sc: SERVER,
		msg: "<target> :Duplicate recipients. No message delivered",
	},
	"409": {
		sc: SERVER,
		msg: ":No origin specified",
	},
	"411": {
		sc: SERVER,
		msg: ":No recipient given %(<command>%)",
	},
	//TODO: Fix below's syntax
	"412": {
		sc: SERVER,
		msg: ":No text to send",
	},
	"413": {
		sc: SERVER,
		msg: "<mask> :No toplevel domain specified",
	},
	"414": {
		sc: SERVER,
		msg: "<mask> :Wildcard in toplevel domain",
	},
	"421": {
		sc: SERVER,
		msg: "<command> :Unknown command",
	},
	"422": {
		sc: SERVER,
		msg: ":MOTD File is missing",
	},
	"423": {
		sc: SERVER,
		msg: "<server> :No administrative info available",
	},
	"424": {
		sc: SERVER,
		msg: ":File error doing <file op> on <file>",
	},
	"431": {
		sc: SERVER,
		msg: ":No nickname given",
	},
	"432": {
		sc: SERVER,
		msg: "<nick> :Erroneus nickname",
	},
	"433": {
		sc: SERVER,
		msg: "<nick> :Nickname is already in use",
	},
	"436": {
		sc: SERVER,
		msg: "<nick> :Nickname collision KILL",
	},
	"441": {
		sc: SERVER,
		msg: "<nick> <channel> :They aren't on that channel",
	},
	"442": {
		sc: SERVER,
		msg: "<channel> :You're not on that channel",
	},
	"443": {
		sc: SERVER,
		msg: "<user> <channel> :is already on channel",
	},
	"444": {
		sc: SERVER,
		msg: "<user> :User not logged in",
	},
	"445": {
		sc: SERVER,
		msg: ":SUMMON has been disabled",
	},
	"446": {
		sc: SERVER,
		msg: ":USERS has been disabled",
	},
	"451": {
		sc: SERVER,
		msg: ":You have not registered",
	},
	"461": {
		sc: SERVER,
		msg: "<command> :Not enough parameters",
	},
	"462": {
		sc: SERVER,
		msg: ":You may not reregister",
	},
	"463": {
		sc: SERVER,
		msg: ":Your host isn't among the privileged",
	},
	"464": {
		sc: SERVER,
		msg: ":Password incorrect",
	},
	"465": {
		sc: SERVER,
		msg: ":You are banned from this server",
	},
	"467": {
		sc: SERVER,
		msg: "<channel> :Channel key already set",
	},
	"471": {
		sc: SERVER,
		msg: "<channel> :Cannot join channel (+l)",
	},
	"472": {
		sc: SERVER,
		msg: "<char> :is unknown mode char to me",
	},
	"473": {
		sc: SERVER,
		msg: "<channel> :Cannot join channel (+i)",
	},
	"474": {
		sc: SERVER,
		msg: "<channel> :Cannot join channel (+b)",
	},
	"475": {
		sc: SERVER,
		msg: "<channel> :Cannot join channel (+k)",
	},
	"481": {
		sc: SERVER,
		msg: ":Permission Denied- You're not an IRC operator",
	},
	"482": {
		sc: SERVER,
		msg: "<channel> :You're not channel operator",
	},
	"483": {
		sc: SERVER,
		msg: ":You cant kill a server!",
	},
	"491": {
		sc: SERVER,
		msg: ":No O-lines for your host",
	},
	"501": {
		sc: SERVER,
		msg: ":Unknown MODE flag",
	},
	"502": {
		sc: SERVER,
		msg: ":Cant change mode for other users",
	},
	//TODO: Fix above's syntax

	"JOIN": {
		sc: SERVER|CLIENT,
		msg: "<,|channel>* <,|key>*",
		rsp: [
			//... a lot
		],
	},
	"KICK": {
		sc: SERVER|CLIENT,
		msg: [
			"<channel> <user>",
			"<channel> <user> :<comment>",
		],
		rsp: [
		],
	},
	"MODE": {
		sc: SERVER|CLIENT,
		//Argument order is limit, users, ban masks
		msg: "<channel> /modes:[+\\-][^ ]+/ <args>*",
		rsp: [
		],
	},
	"NICK": {
		sc: SERVER|CLIENT,
		msg: ["<nickname>","<nickname> <hopcount>"],
		split: ["support",splitWholeEntries],
		rsp: ["431","432","433","436"],
	},
	"NOTICE": {
		sc: SERVER|CLIENT,
		msg: "<target> :{text}",
		split: ["text",splitFormattedText],
	},
	"OPER": {
		sc: CLIENT,
		msg: "<name> <password>",
	},
	"PART": {
		sc: CLIENT|SERVER,
		msg: "<,|channel>*"
	},
	"PASS": {
		sc: CLIENT,
		msg: "<password>",
		rsp: ["461","462"],
	},
	"PING": {
		sc: SERVER|CLIENT,
		msg: "<server1> <server2>",
		rsp: ["PONG",//...
		],
	},
	"PONG": {
		sc: SERVER|CLIENT,
		msg: "<daemon1> <daemon2>",
		rsp: [
		],
	},
	"PRIVMSG": {
		sc: SERVER|CLIENT,
		msg: "<target> :{text}",
		split: ["text",splitFormattedText],
	},
	"QUIT": {
		sc: SERVER|CLIENT,
		msg: ":<message>",
	},
	"USER": {
		sc: CLIENT,
		rfc1459: "<username> <hostname> <servername> <realname>",
		rfc281x: "<username> <mode> * :{realname}",
		rsp: ["461","462"],
	},
}

//Make all splitters
for(var x in msgs){
	var msg = msgs[x];
	if(Array.isArray(msg)){
		for(var i=0;i<msg.length;++i){
			makeSplitter(msg,i,x.charAt(0) == "_");
		}
	} else makeSplitter(msgs,x,x.charAt(0) == "_");
}

function makeSplitter(msgs,x,mode){
	var msg = msgs[x];
	if(mode){
		msgs[x] = new Splitter(msg);
	} else {
		if("msg" in msg){
			msg.msg = new Splitter(msg.msg);
		}
		for(var i=0;i<2;i++){
			var rfc = i==0?"rfc1459":"rfc281x";
			if(rfc in msg){
				if(typeof(msg[rfc]) == "string"){
					msg[rfc] = {
						msg: new Splitter(msg[rfc])
					}
				} else if("msg" in msg[rfc]) {
					msg[rfc].msg = new Splitter(msg[rfc].msg);
				}
			}
			if(!("split" in msg)){
				msg.split = null;
			}
		}
	}
}
