// irc/proto.js
/*
Describe messages sent by the server and how to read them
:sender command params\r\n
command is the key, params described in the value (bounds are implied)
Everything's literally matched except:
{name} matches a value potentially with spaces and saves to name (.*)
{name:N} same as above, but may only contain up to N spaces
{name=N} same as above, but must contain exactly N spaces
{name?cond} matches a values with spaces only if the condition is met,
+otherwise it matches a value without spaces.
+Cond should be a string, by default it is checked to be in existence at the
+beginning of the match. Cond can be prefixed with a ? to show that the spaced
+match is optional (ie. that matching only cond is possible). Prefixing cond
+with > says that cond should be matched on the right of the message instead.
+Similarly, prefixing it with < ensures that it matches on the left. < or >
+come after the ?
<name> matches a value without spaces and saves to name ([^ ]*)
<name:etc> same as above, but characters in etc are also not allowed
+Note that to use a literal - instead of a range, use --
<name=etc> match a value that only contains etc. Put a + on the end
+to match multiple of those characters.
<name!etc> same as above but negative.
<#name> matches number and saves to name ([0-9]+)
<#name:.,> matches number with given modifiers ([0-9,]+(\.[0-9]+)?)
+Put an s on the end to switch the meaning of . and ,
<name@otherName> inherets the match rules of otherName. Preceding otherName
+with a _ references the global rule, otherwise references one in the
+current message. Naming is nested for globals.
/name:regex/ specific match (grouping is implied) note: regex is JS's
suffixing with a * (ie. directly after } > or /) will cause the same
+thing to match multiple times and return an array under that name
escape {}<>/%* by preceeding it with a %
Can be an array for multiple forms.
Sections with the same name will assert equivalence within that key
You may denote a section with ()
Sections surrounded by [] denote optional sections.
If there are multiple options per section use [grp1]|[grp2]
If they're not optional, use (grp1)|(grp2)
For all groups - {} <> // () [] - you may start it with one character
+followed by a | to denote what the separator is between multiple
+instances of this section. By default it's space
*/

/*
Message Object {
	sc: S2C | C2S, //sent from Server and/or Client
	//Can be arrays
	msg: "", //Message passed to splitter, default
	rsp: [], //Possible response commands, default
	//The below can also be strings, assumed to be msg
	rfc1459 { //only using rfc1459 (this before default)
		msg: "",
		rsp: [],
	},
	rfc281x: { //with rfc2810-rfc2813 enabled
		msg: "",
		rsp: [],
	},
}
*/

var Splitter = require("splitter.js");

//Server to Client, Client to Server
var S2C=1,C2S=2,INTERSERVER=4,msgs;

function splitWholeEntries(len,data){
	var ret=[];
	while(data.length){
		var rlen = -1;
		for(var i=0;i<data.length && (rlen+=(data[i].length+1)) <= len;++i);
		if(i == 0)
		ret.push(data.slice(0,i));
		data = data.slice(i);
	}
	return ret;
}

function splitFormattedText(len,data){
	var ret = [];
	while(data.length){
		//Make sure to carry over formatting and CTCPs if they can't be completed on that line
		//Word wrap, as well
		//Reference: CTCP \1, Bold \2, Color \3, Italic/Reversed \26, Underline \37, Unformat \17
		var rstr = data.substr(0,len)
		,   rmdr = data.substr(len)
		,   pos,fmt;
		if(rstr.match(/\1/g).length % 2 == 1){
			//CTCP was cut off
			pos = rstr.lastIndexOf("\1");
			rmdr = rstr.substr(pos)+rmdr;
			rstr = rstr.substr(0,pos);
		}
		//Split to last bound (if it's not too far)
		var tmp;
		if(tmp=rstr.match(/(.*[^a-zA-Z0-9\1\2\3\26\37\17].*?)([^\1]*)$/)
		&& tmp[2].length < 20 && !rmdr.match(/^[^a-zA-Z0-9\1\2\3\26\37\17]/)){
			rstr = tmp[1];
			rmdr = tmp[2]+rmdr;
		}
		fmt = rstr.match(/[\2\3\26\37\17]/g);
		if((pos = fmt.lastIndexOf("\17")) != -1){
			//Stopper was found, only retain what comes after
			fmt.slice(pos+1);
		}
		if(fmt.length){
			//Formatting left over, prepend to remainder
			rmdr = fmt.join("")+rmdr;
		}

		ret.push(rstr)
		data = rmdr;
	}
	return ret;
}

msgs = {
	"_server": ":<sender> <command>[ {args}]",
	"_client": "<command>[ {args}]",
	"_numeric": "<nick>[ {args}]",
	//"_args": "<spaced>* :{trailing}",
	"_target": "(,|"+
		"(/prefix:[#+&]|![A-Z0-9]{5}/<channel:\0\7\r\n,:>[:<mask@channel>])|"+
		"(<user>[%<host>]@<server>)|"+
		"(<user>%<host>)|"+
		"(<nick>[!<user>@<host>])|"+
		//"([$#]"
	")*",
	"_hostname": "<.|@_shortname>*",
	"_shortname": "/:[a-zA-Z0-9][a-zA-Z0-9--][a-zA-Z0-9]/",
	
	//TODO: Remove specific messages, add them to an english language file for them
	
	//Format for key is command[-<desc:->][-<conflictIndex:->]
	//Format for data is [direction, message, [split, splitFunction]] split section is optional
	"001-RPL_WELCOME": [S2C,":Welcome to the {network} <nick>!<user>@<host>"],
	"002-RPL_YOURHOST": [S2C,":Your host is <serverName>, running version <version>"],
	"003-RPL_CREATED": [S2C,":This server was created {date}"],
	"004-RPL_MYINFO-1": [S2C,"<serverName> <version> <okUserModes> <okChannelModes>"],
	"004-RPL_MYINFO-2": [S2C,"<serverName> <version> <okUserModes> <okChannelModes>"
		+" <paramChannelModes> <paramUserModes> <okServerModes> <paramServerModes>"
	],
	"005-RPL_ISUPPORT": [S2C,"<support>* :are supported by this server",
		["support",splitWholeEntries],
	],
	"005-RPL_BOUNCE": [S2C,":Try server <server>, port <#port>"],
	//"006-RPL_MAP"
	//"007-RPL_MAPEND"
	//"008-RPL_SNOMASK"
	//"009-RPL_STATMEMTOT"
	"010-RPL_BOUNCE": [S2C,"<host> <#port> :{info}"],
	//"010-RPL_STATMEM" //deprecated
	
	//"014-RPL_YOURCOOKIE"
	//"015-RPL_MAP"
	//"016-RPL_MAPMORE"
	//"017-RPL_MAPEND"
	
	//"042-RPL_YOURID"
	//"043-RPL_SAVENICK"
	
	//"050-RPL_ATTEMPTINGJUNC"
	//"051-RPL_ATTEMPTINGREROUTE"
	
	//NOTE: Correct?
	"200-RPL_TRACELINK-1": [S2C,"Link <version>[.<debugLevel>] <destination> <nextServer>"],
	"200-RPL_TRACELINK-2": [S2C,"Link <version>[.<debugLevel>] <destination> <nextServer>"
		+" V<protocolVersion> <linkUptimeSec> <backstreamSendq> <upstreamSendq>"
	],
	"201-RPL_TRACECONNECTING": [S2C,"Try. <class> <server>"],
	"202-RPL_TRACEHANDSHAKE": [S2C,"H.S. <class> <server>"],
	"203-RPL_TRACEUNKNOWN": [S2C,"???? <class>[ <clientIP>]"],
	"204-RPL_TRACEOPERATOR": [S2C,"Oper <class> <nick>"],
	"205-RPL_TRACEUSER": [S2C,"User <class> <nick>"],
	"206-RPL_TRACESERVER-1": [S2C,"Serv <class> <#S>S <#C>C <server> <nick>!<user>@<host>"],
	"206-RPL_TRACESERVER-2": [S2C,"Serv <class> <#S>S <#C>C <server> <nick>!<user>@<host> V<protocolVersion>"],
	"207-RPL_TRACESERVICE": [S2C,"Service <class> <name> <type> <activeType>"],
	"208-RPL_TRACENEWTYPE": [S2C,"<newtype> 0 <clientName>"],
	"209-RPL_TRACECLASS": [S2C,"Class <class> <count>"],
	//"210-RPL_TRACERECONNECT" //deprecated
	//"210-RPL_STATS"
	"211-RPL_STATSLINKINFO": [S2C,"<linkname> <sendq> <sentMessages> <sentBytes> <receivedMessages> <receivedBytes> <timeOpen>"],
	"212-RPL_STATSCOMMANDS": [S2C,"<command> <count>"],
	"213-RPL_STATSCLINE": [S2C,"C <host> * <name> <port> <class>"],
	"214-RPL_STATSNLINE": [S2C,"N <host> * <name> <port> <class>"],
	"215-RPL_STATSILINE": [S2C,"I <host> * <host> <port> <class>"],
	"216-RPL_STATSKLINE": [S2C,"K <host> * <username> <port> <class>"],
	//"217-RPL_STATSQLINE"
	//"217-RPL_STATSPLINE"
	"218-RPL_STATSYLINE": [S2C,"Y <class> <pingFrequency> <connectFrequency> <maxSendq>"],
	"219-RPL_ENDOFSTATS": [S2C,"<letter> :End of /STATS report"],
	//"220-RPL_STATSPLINE"
	//"220-RPL_STATSBLINE"
	"221-RPL_UMODEIS": [S2C,"<userModes>"],
	//"222-RPL_MODLIST"
	//"222-RPL_SQLINE_NICK"
	//"222-RPL_STATSBLINE"
	//"223-RPL_STATSGLINE"
	//"224-RPL_STATSFLINE"
	//"224-RPL_STATSTLINE"
	//"225-RPL_STATSDLINE"
	//"225-RPL_STATSZLINE"
	//"225-RPL_STATSELINE"
	//"226-RPL_STATSCOUNT"
	//"226-RPL_STATSNLINE"
	//"227-RPL_STATSGLINE"
	//"227-RPL_STATSVLINE"
	//"228-RPL_STATSQLINE"
	
	//"231-RPL_SERVICEINFO" //deprecated
	//"232-RPL_ENDOFSERVICES" //deprecated
	//"232-RPL_RULES"
	//"233-RPL_SERVICE" //deprecated
	//"234-RPL_SERVLIST"
	//"235-RPL_SERVLISTEND"
	//"236-RPL_STATSVERBOSE"
	//"237-RPL_STATSENGINE"
	//"238-RPL_STATSFLINE"
	//"239-RPL_STATSIAUTH"
	//"240-RPL_STATSVLINE"
	//"240-RPL_STATSXLINE"
	"241-RPL_STATSLLINE": [S2C,"L <hostmask> * <servername> <maxdepth>"],
	"242-RPL_STATSUPTIME": [S2C,":Server Up <#days> days <#hour>:<#min>:<#sec>"],
	"243-RPL_STATSOLINE": [S2C,"O <hostmask> * <name>"],
	"244-RPL_STATSHLINE": [S2C,"H <hostmask> * <servername>"],
	//"245-RPL_STATSSLINE"
	//"246-RPL_STATSPING"
	//"246-RPL_STATSTLINE"
	//"246-RPL_STATSULINE"
	//"247-RPL_STATSBLINE"
	//"247-RPL_STATSXLINE"
	//"247-RPL_STATSGLINE"
	//"248-RPL_STATSULINE"
	//"248-RPL_STATSDEFINE"
	//"249-RPL_STATSULINE"
	//"249-RPL_STATSDEBUG"
	//"250-RPL_STATSDLINE"
	//"250-RPL_STATSCONN"
	"251-RPL_LUSERCLIENT": [S2C,":There are <#userCount> users and <#invisCount> invisible on <#servCount> servers"],
	"252-RPL_LUSEROP": [S2C,"<#opsCount> :operator%(s%) online"],
	"253-RPL_LUSERUNKNOWN": [S2C,"<#unknownConns> :unknown connection%(s%)"],
	"254-RPL_LUSERCHANNELS": [S2C,"<#chanCount> :channels formed"],
	"255-RPL_LUSERME": [S2C,":I have <#clients> clients and <#servers> servers"],
	"256-RPL_ADMINME-1": [S2C,"<server> :Administrative info"],
	"256-RPL_ADMINME-2": [S2C,"[<server> ]:Administrative info about <server>"],
	"257-RPL_ADMINLOC1": [S2C,":{adminLocation1}?"],
	"258-RPL_ADMINLOC2": [S2C,":{adminLocation2}?"],
	"259-RPL_ADMINEMAIL": [S2C,":{adminEmail}?"],
	
	"261-RPL_TRACELOG": [S2C,"File <logfile> <debuglevel>"],
	//"262-RPL_TRACEPING"
	//"262-RPL_TRACEEND"
	//"263-RPL_TRYAGAIN"
	
	//"265-RPL_LOCALUSERS"
	//"266-RPL_GLOBALUSERS"
	//"267-RPL_START_NETSTAT"
	//"268-RPL_NETSTAT"
	//"269-RPL_END_NETSTAT"
	//"270-RPL_PRIVS"
	//"271-RPL_SILELIST"
	//"272-RPL_ENDOFSILELIST"
	//"273-RPL_NOTIFY"
	//"274-RPL_ENDNOTIFY"
	//"274-RPL_STATSDELTA"
	//"275-RPL_STATSDLINE"
	//"276-RPL_VCHANEXIST"
	//"277-RPL_VCHANLIST"
	//"278-RPL_VCHANHELP"
	
	//"280-RPL_GLIST"
	//"281-RPL_ENDOFGLIST"
	//"281-RPL_ACCEPTLIST"
	//"282-RPL_ENDOFACCEPT"
	//"282-RPL_JUPELIST"
	//"283-RPL_ALIST"
	//"283-RPL_ENDOFJUPELIST"
	//"284-RPL_ENDOFALIST"
	//"284-RPL_FEATURE"
	//"285-RPL_GLIST_HASH"
	//"285-RPL_CHANINFO_HANDLE"
	//"285-RPL_NEWHOSTIS"
	//"286-RPL_CHANINFO_USERS"
	//"286-RPL_CHKHEAD"
	//"287-RPL_CHANINFO_CHOPS"
	//"287-RPL_CHANUSER"
	//"288-RPL_CHANINFO_VOICES"
	//"288-RPL_PATCHHEAD"
	//"289-RPL_CHANINFO_AWAY"
	//"289-RPL_PATCHCON"
	//"290-RPL_CHANINFO_OPERS"
	//"290-RPL_HELPHDR"
	//"290-RPL_DATASTR"
	//"291-RPL_CHANINFO_BANNED"
	//"291-RPL_HELPOP"
	//"291-RPL_ENDOFCHECK"
	//"292-RPL_CHANINFO_BANS"
	//"292-RPL_HELPTLR"
	//"293-RPL_CHANINFO_INVITE"
	//"293-RPL_HELPHLP"
	//"294-RPL_CHANINFO_INVITES"
	//"294-RPL_HELPFWD"
	//"295-RPL_CHANINFO_KICK"
	//"295-RPL_HELPIGN"
	//"296-RPL_CHANINFO_KICKS"
	
	//"299-RPL_END_CHANINFO"
	"300-RPL_NONE": [S2C,"[{data}]"],
	"301-RPL_AWAY": [S2C,"<nick> :{awayMessage}"],
	"302-RPL_USERHOST": [S2C,":[<nick:*=>/op:%*?/=<away=+-><host>]*"],
	"303-RPL_ISON": [S2C,":<nick>*?"],
	//"304-RPL_TEXT" //deprecated
	"305-RPL_UNAWAY": [S2C,":You are no longer marked as being away"],
	"306-RPL_NOWAWAY": [S2C,":You have been marked as being away"],
	//"307-RPL_USERIP"
	//"307-RPL_WHOISREGNICK"
	//"307-RPL_SUSERHOST"
	//"308-RPL_NOTIFYACTION"
	//"308-RPL_WHOISADMIN"
	//"308-RPL_RULESSTART"
	//"309-RPL_NICKTRACE"
	//"309-RPL_WHOISSADMIN"
	//"309-RPL_ENDOFRULES"
	//"309-RPL_WHOISHELPER"
	//"310-RPL_WHOISSVCMSG"
	//"310-RPL_WHOISHELPOP"
	//"310-RPL_WHOISSERVICE"
	"311-RPL_WHOISUSER": [S2C,"<nick> <user> <host> * :{realname}"],
	"312-RPL_WHOISSERVER": [S2C,"<nick> <server> :{serverInfo}"],
	"313-RPL_WHOISOPERATOR": [S2C,"<nick> :is an IRC operator"],
	"314-RPL_WHOWASUSER": [S2C,"<nick> <user> <host> * :{realname}"],
	"315-RPL_ENDOFWHO": [S2C,"<name> :End of /WHO list"],
	//"316-RPL_WHOISCHANOP" //deprecated
	"317-RPL_WHOISIDLE": [S2C,"<nick> <#seconds> :seconds idle"],
	"318-RPL_ENDOFWHOIS": [S2C,"<nick> :End of /WHOIS list"],
	"319-RPL_WHOISCHANNELS": [S2C,"<nick> :<channel>*"],
	//"320-RPL_WHOISVIRT"
	//"320-RPL_WHOIS_HIDDEN"
	//"320-RPL_WHOISSPECIAL"
	"321-RPL_LISTSTART": [S2C,"Channel :Users<= +>Name"], //deprecated
	"322-RPL_LIST": [S2C,"<channel> <#userCount> :{topic}"],
	"323-RPL_LISTEND": [S2C,":End of /LIST"],
	"324-RPL_CHANNELMODEIS": [S2C,"<channel> <modes> <modeParams>*"],
	//"325-RPL_UNIQOPIS"
	//"325-RPL_CHANNELPASSIS"
	//"326-RPL_NOCHANPASS"
	//"327-RPL_CHPASSUNKNOWN"
	//"328-RPL_CHANNEL_URL"
	//"329-RPL_CREATIONTIME"
	//"330-RPL_WHOWAS_TIME"
	//"330-RPL_WHOISACCOUNT"
	"331-RPL_NOTOPIC": [S2C,"<channel> :No topic is set"],
	"332-RPL_TOPIC": [S2C,"<channel> :<topic>"],
	//"333-RPL_TOPICWHOTIME"
	//"334-RPL_LISTUSAGE"
	//"334-RPL_COMMANDSYNTAX"
	//"334-RPL_LISTSYNTAX"
	//"335-RPL_WHOISBOT"
	
	//"338-RPL_CHANPASSOK"
	//"338-RPL_WHOISACTUALLY"
	//"339-RPL_BADCHANPASS"
	//"340-RPL_USERIP"
	//NOTE: Use <channel> <nick> in rfc1459 strictness mode, but <nick> <channel> otherwise
	"341-RPL_INVITING-1": [S2C,"/channel:[@#+!].*/ <nick>"],
	"341-RPL_INVITING-2": [S2C,"<nick> /channel:[@#+!].*/"],
	"342-RPL_SUMMONING": [S2C,"<user> :Summoning user to IRC"],
	
	//"345-RPL_INVITED"
	//"346-RPL_INVITELIST"
	//"347-RPL_ENDOFINVITELIST"
	//"348-RPL_EXCEPTLIST"
	//"349-RPL_ENDOFEXCEPTLIST"
	
	"351-RPL_VERSION": [S2C,"<version>[.<debuglevel>] <server> :{comments}"],
	//TODO: What does the asterisk mean? I see an r sometimes, is that the status? What is HG?
	"352-RPL_WHOREPLY": [S2C,"<channel> <user> <host> <server> <nick> <HG=HG>/asterisk:*//status:.?/ :<hopcount> {realName}"],
	"353-RPL_NAMREPLY": [S2C,"<channel> :<nicks>*"],
	//"354-RPL_WHOSPCRPL"
	//"355-RPL_NAMREPLY_"
	
	//"357-RPL_MAP"
	//"358-RPL_MAPMORE"
	//"359-RPL_MAPEND"
	
	//"361-RPL_KILLDONE" //deprecated
	//"362-RPL_CLOSING" //deprecated
	//"363-RPL_CLOSEEND" //deprecated
	"364-RPL_LINKS": [S2C,"<mask> <server> :<hopcount> {serverInfo}"],
	"365-RPL_ENDOFLINKS": [S2C,"<mask> :End of /LINKS list"],
	"366-RPL_ENDOFNAMES": [S2C,"<channel> :End of /NAMES list"],
	"367-RPL_BANLIST": [S2C,"<channel> <banid>"],
	"368-RPL_ENDOFBANLIST": [S2C,"<channel> :End of channel ban list"],
	"369-RPL_ENDOFWHOWAS": [S2C,"<nick> :End of WHOWAS"],
	
	"371-RPL_INFO": [S2C,":{info}"],
	"372-RPL_MOTD": [S2C,":-[ {motd}]"],
	//"373-RPL_INFOSTART" //deprecated
	"374-RPL_ENDOFINFO": [S2C,":End of /INFO list"],
	"375-RPL_MOTDSTART": [S2C,":- <server> Message of the day - "],
	"376-RPL_ENDOFMOTD": [S2C,":End of /MOTD command"],
	//"377-RPL_KICKEXPIRED" //deprecated
	//"377-RPL_SPAM"
	//"378-RPL_BANEXPIRED"
	//"378-RPL_WHOISHOST"
	//"378-RPL_MOTD" //deprecated
	//"379-RPL_KICKLINKED"
	//"379-RPL_WHOISMODES"
	//"380-RPL_BANLINKED"
	//"380-RPL_YOURHELPER"
	"381-RPL_YOUREOPER": [S2C,":You are now an IRC operator"],
	"382-RPL_REHASHING": [S2C,"<configFile> :Rehashing"],
	//"383-RPL_YOURESERVICE"
	//"384-RPL_MYPORTIS" //deprecated
	//"385-RPL_NOTOPERANYMORE"
	//"386-RPL_QLIST"
	//"386-RPL_IRCOPS"
	//"387-RPL_ENDOFQLIST"
	//"387-RPL_ENDOFIRCOPS"
	//"388-RPL_ALIST"
	//"389-RPL_ENDOFALIST"
	
	"391-RPL_TIME-1": [S2C,"<server> :{timeString}"],
	"391-RPL_TIME-2": [S2C,"<server> <#timestamp> <#offset> :{timeString}"],
	"391-RPL_TIME-3": [S2C,"<server> <timezone> <#microseconds> :{timeString}"],
	"391-RPL_TIME-4": [S2C,"<server> <#year> <#month> <#day> <#hour> <#minute> #<second>"],
	"392-RPL_USERSSTART": [S2C,":UserID Terminal Host"],
	"393-RPL_USERS": [S2C,":<username> <ttyline> <hostname>"],
	"394-RPL_ENDOFUSERS": [S2C,":End of users"],
	"395-RPL_NOUSERS": [S2C,":Nobody logged in"],
	//"396-RPL_HOSTHIDDEN"
	
	//"400-ERR_UNKNOWNERROR"
	"401-ERR_NOSUCHNICK": [S2C,"<nick> :No such nick/channel"],
	"402-ERR_NOSUCHSERVER": [S2C,"<server> :No such server"],
	"403-ERR_NOSUCHCHANNEL": [S2C,"<channel> :No such channel"],
	"404-ERR_CANNOTSENDTOCHAN": [S2C,"<channel> :Cannot send to channel"],
	"405-ERR_TOOMANYCHANNELS": [S2C,"<channel> :You have joined too many channels"],
	"406-ERR_WASNOSUCHNICK": [S2C,"<nick> :There was no such nickname"],
	"407-ERR_TOOMANYTARGETS": [S2C,"<target> :Duplicate recipients. No message delivered"],
	//"408-ERR_NOSUCHSERVICE"
	//"408-ERR_NOCOLORSONCHAN"
	"409-ERR_NOORIGIN": [S2C,":No origin specified"],
	
	"411-ERR_NORECIPIENT": [S2C,":No recipient given %(<command>%)"],
	"412-ERR_NOTEXTTOSEND": [S2C,":No text to send"],
	"413-ERR_NOTOPLEVEL": [S2C,"<mask> :No toplevel domain specified"],
	"414-ERR_WILDTOPLEVEL": [S2C,"<mask> :Wildcard in toplevel domain"],
	//"415-ERR_BADMASK"
	//"416-ERR_TOOMANYMATCHES"
	//"416-ERR_QUERYTOOLONG"
	
	//"419-ERR_LENGTHTRUNCATED"
	
	"421-ERR_UNKNOWNCOMMAND": [S2C,"<command> :Unknown command"],
	"422-ERR_NOMOTD": [S2C,":MOTD File is missing"],
	"423-ERR_NOADMININFO": [S2C,"<server> :No administrative info available"],
	"424-ERR_FILEERROR": [S2C,":{reason}"],
	//"425-ERR_NOOPERMOTD"
	
	//"429-ERR_TOOMANYAWAY"
	//"430-ERR_EVENTNICKCHANGE"
	"431-ERR_NONICKNAMEGIVEN": [S2C,":No nickname given"],
	"432-ERR_ERRONEUSNICKNAME": [S2C,"<nick> :Erroneus nickname"],
	"433-ERR_NICKNAMEINUSE": [S2C,"<nick> :Nickname is already in use"],
	//"434-ERR_SERVICENAMEINUSE"
	//"434-ERR_NORULES"
	//"435-ERR_SERVICECONFUSED"
	//"435-ERR_BANONCHAN"
	"436-ERR_NICKCOLLISION": [S2C,"<nick> :Nickname collision KILL"],
	//"437-ERR_UNAVAILRESOURCE"
	//"437-ERR_BANNICKCHANGE"
	//"438-ERR_NICKTOOFAST"
	//"438-ERR_DEAD"
	//"439-ERR_TARGETTOOFAST"
	//"440-ERR_SERVICESDOWN"
	"441-ERR_USERNOTINCHANNEL": [S2C,"<nick> <channel> :They aren't on that channel"],
	"442-ERR_NOTONCHANNEL": [S2C,"<channel> :You're not on that channel"],
	"443-ERR_USERONCHANNEL": [S2C,"<nick> <channel> :is already on channel"],
	"444-ERR_NOLOGIN": [S2C,"<user> :User not logged in"],
	"445-ERR_SUMMONDISABLED": [S2C,":SUMMON has been disabled"],
	"446-ERR_USERSDISABLED": [S2C,":USERS has been disabled"],
	//"447-ERR_NONICKCHANGE"
	
	//"449-ERR_NOTIMPLEMENTED"
	
	"451-ERR_NOTREGISTERED": [S2C,":You have not registered"],
	//"452-ERR_IDCOLLISION"
	//"453-ERR_NICKLOST"
	
	//"455-ERR_HOSTILENAME"
	//"456-ERR_ACCEPTFULL"
	//"457-ERR_ACCEPTEXIST"
	//"458-ERR_ACCEPTNOT"
	//"459-ERR_NOHIDING"
	//"460-ERR_NOTFORHALFOPS"
	"461-ERR_NEEDMOREPARAMS": [S2C,"<command> :Not enough parameters"],
	"462-ERR_ALREADYREGISTRED": [S2C,":You may not reregister"],
	"463-ERR_NOPERMFORHOST": [S2C,":Your host isn't among the privileged"],
	"464-ERR_PASSWDMISMATCH": [S2C,":Password incorrect"],
	"465-ERR_YOUREBANNEDCREEP": [S2C,":You are banned from this server"],
	//"466-ERR_YOUWILLBEBANNED" //deprecated
	"467-ERR_KEYSET": [S2C,"<channel> :Channel key already set"],
	//"468-ERR_INVALIDUSERNAME"
	//"468-ERR_ONLYSERVERSCANCHANGE"
	//"469-ERR_LINKSET"
	//"470-ERR_LINKCHANNEL"
	//"470-ERR_KICKEDFROMCHAN"
	"471-ERR_CHANNELISFULL": [S2C,"<channel> :Cannot join channel %(+l%)"],
	"472-ERR_UNKNOWNMODE": [S2C,"<char> :is unknown mode char to me"],
	"473-ERR_INVITEONLYCHAN": [S2C,"<channel> :Cannot join channel %(+i%)"],
	"474-ERR_BANNEDFROMCHAN": [S2C,"<channel> :Cannot join channel %(+b%)"],
	"475-ERR_BADCHANNELKEY": [S2C,"<channel> :Cannot join channel %(+k%)"],
	//"476-ERR_BADCHANMASK"
	//"477-ERR_NOCHANMODES"
	//"477-ERR_NEEDREGGEDNICK"
	//"478-ERR_BANLISTFULL"
	//"479-ERR_BADCHANNAME"
	//"479-ERR_LINKFAIL"
	//"480-ERR_NOULINE"
	//"480-ERR_CANNOTKNOCK"
	"481-ERR_NOPRIVILEGES": [S2C,":Permission Denied- You're not an IRC operator"],
	"482-ERR_CHANOPRIVSNEEDED": [S2C,"<channel> :You're not channel operator"],
	"483-ERR_CANTKILLSERVER": [S2C,":You cant kill a server!"],
	//"484-ERR_RESTRICTED"
	//"484-ERR_ISCHANSERVICE"
	//"484-ERR_DESYNC"
	//"484-ERR_ATTACKDENY"
	//"485-ERR_UNIQOPRIVSNEEDED"
	//"485-ERR_KILLDENY"
	//"485-ERR_CANTKICKADMIN"
	//"485-ERR_ISREALSERVICE"
	//"486-ERR_NONONREG"
	//"486-ERR_HTMDISABLED"
	//"486-ERR_ACCOUNTONLY"
	//"487-ERR_CHANTOORECENT"
	//"487-ERR_MSGSERVICES"
	//"488-ERR_TSLESSCHAN"
	//"489-ERR_VOICENEEDED"
	//"489-ERR_SECUREONLYCHAN"
	
	"491-ERR_NOOPERHOST": [S2C,":No O-lines for your host"],
	//"492-ERR_NOSERVICEHOST" //deprecated
	//"493-ERR_NOFEATURE"
	//"494-ERR_BADFEATURE"
	//"495-ERR_BADLOGTYPE"
	//"496-ERR_BADLOGSYS"
	//"497-ERR_BADLOGVALUE"
	//"498-ERR_ISOPERLCHAN"
	//"499-ERR_CHANOWNPRIVNEEDED"
	
	"501-ERR_UMODEUNKNOWNFLAG": [S2C,":Unknown MODE flag"],
	"502-ERR_USERSDONTMATCH": [S2C,":Cant change mode for other users"],
	//"503-ERR_VWORLDWARN"
	//"504-ERR_USERNOTONSERV"
	
	//"511-ERR_SILELISTFULL"
	//"512-ERR_TOOMANYWATCH"
	//"513-ERR_BADPING"
	//"514-ERR_INVALID_ERROR"
	//"514-ERR_TOOMANYDCC"
	//"515-ERR_BADEXPIRE"
	//"516-ERR_DONTCHEAT"
	//"517-ERR_DISABLED"
	//"518-ERR_NOINVITE"
	//"518-ERR_LONGMASK"
	//"519-ERR_ADMONLY"
	//"519-ERR_TOOMANYUSERS"
	//"520-ERR_OPERONLY"
	//"520-ERR_MASKTOOWIDE"
	//"520-ERR_WHOTRUNC"
	//"521-ERR_LISTSYNTAX"
	//"522-ERR_WHOSYNTAX"
	//"523-ERR_WHOLIMEXCEED"
	//"524-ERR_QUARANTINED"
	//"524-ERR_OPERSPVERIFY"
	//525-526 - http://bgp.potaroo.net/ietf/all-ids/draft-brocklesby-irc-usercmdpfx-00.txt
	//"525-ERR_REMOTEPFX"
	//"526-ERR_PFXUNROUTABLE"
	
	//"550-ERR_BADHOSTMASK"
	//"551-ERR_HOSTUNAVAIL"
	//"552-ERR_USINGSLINE"
	//"553-ERR_STATSSLINE"
	
	//"600-RPL_LOGON"
	//"601-RPL_LOGOFF"
	//"602-RPL_WATCHOFF"
	//"603-RPL_WATCHSTAT"
	//"604-RPL_NOWON"
	//"605-RPL_NOWOFF"
	//"606-RPL_WATCHLIST"
	//"607-RPL_ENDOFWATCHLIST"
	//"608-RPL_WATCHCLEAR"
	
	//"610-RPL_MAPMORE"
	//"610-RPL_ISOPER"
	//"611-RPL_ISLOCOP"
	//"612-RPL_ISNOTOPER"
	//"613-RPL_ENDOFISOPER"
	
	//"615-RPL_MAPMORE"
	//"615-RPL_WHOISMODES"
	//"616-RPL_WHOISHOST"
	//"617-RPL_DCCSTATUS"
	//"617-RPL_WHOISBOT"
	//"618-RPL_DCCLIST"
	//"619-RPL_ENDOFDCCLIST"
	//"619-RPL_WHOWASHOST"
	//"620-RPL_DCCINFO"
	//"620-RPL_RULESSTART"
	//"621-RPL_RULES"
	//"622-RPL_ENDOFRULES"
	//"623-RPL_MAPMORE"
	//"624-RPL_OMOTDSTART"
	//"625-RPL_OMOTD"
	//"626-RPL_ENDOFO"
	
	//"630-RPL_SETTINGS"
	//"631-RPL_ENDOFSETTINGS"
	
	//"640-RPL_DUMPING"
	//"641-RPL_DUMPRPL"
	//"642-RPL_EODUMP"
	
	//"660-RPL_TRACEROUTE_HOP"
	//"661-RPL_TRACEROUTE_START"
	//"662-RPL_MODECHANGEWARN"
	//"663-RPL_CHANREDIR"
	//"664-RPL_SERVMODEIS"
	//"665-RPL_OTHERUMODEIS"
	//"666-RPL_ENDOF_GENERIC"
	
	//"670-RPL_WHOWASDETAILS"
	//"671-RPL_WHOISSECURE"
	//"672-RPL_UNKNOWNMODES"
	//"673-RPL_CANNOTSETMODES"
	
	//"678-RPL_LUSERSTAFF"
	//"679-RPL_TIMEONSERVERIS"
	
	//"682-RPL_NETWORKS"
	
	//"687-RPL_YOURLANGUAGEIS"
	//"688-RPL_LANGUAGE"
	//"689-RPL_WHOISSTAFF"
	//"690-RPL_WHOISLANGUAGE"
	
	//"702-RPL_MODLIST"
	//"703-RPL_ENDOFMODLIST"
	//"704-RPL_HELPSTART"
	//"705-RPL_HELPTXT"
	//"706-RPL_ENDOFHELP"
	
	//"708-RPL_ETRACEFULL"
	//"709-RPL_ETRACE"
	//"710-RPL_KNOCK"
	//"711-RPL_KNOCKDLVR"
	//"712-ERR_TOOMANYKNOCK"
	//"713-ERR_CHANOPEN"
	//"714-ERR_KNOCKONCHAN"
	//"715-ERR_KNOCKDISABLED"
	//"716-RPL_TARGUMODEG"
	//"717-RPL_TARGNOTIFY"
	//"718-RPL_UMODEGMSG"
	
	//"720-RPL_OMOTDSTART"
	//"721-RPL_OMOTD"
	//"722-RPL_ENDOFOMOTD"
	//"723-ERR_NOPRIVS"
	//"724-RPL_TESTMARK"
	//"725-RPL_TESTLINE"
	//"726-RPL_NOTESTLINE"
	
	//770-774 - http://ithildin.org/trac/browser/trunk/ithildin/modules/ircd/commands/xinfo.c
	//"770-ERR_AMBIGUOUSXINFO"
	//"771-RPL_XINFO"
	//"773-RPL_XINFOSTART"
	//"774-RPL_XINFOEND"
	
	//"972-ERR_CANNOTDOCOMMAND"
	//"973-ERR_CANNOTCHANGEUMODE"
	//"974-ERR_CANNOTCHANGECHANMODE"
	//"975-ERR_CANNOTCHANGESERVERMODE"
	//"976-ERR_CANNOTSENDTONICK"
	//"977-ERR_UNKNOWNSERVERMODE"
	
	//"979-ERR_SERVERMODELOCK"
	//"980-ERR_BADCHARENCODING"
	//"981-ERR_TOOMANYLANGUAGES"
	//"982-ERR_NOLANGUAGE"
	//"983-ERR_TEXTTOOSHORT"
	
	//"999-ERR_NUMERIC_ERR"
	
	//NOTE: All word commands should be uppercase
	//NOTE: Keep them in alphabetical order, too

	"ADMIN": [C2S|INTERSERVER,"<server>?"],
	"AWAY": [C2S,"[:{message}]"],
	"CONNECT": [C2S|INTERSERVER,"<targetServer>[ <#port>[ <remoteServer>]]"],
	"ERROR": [INTERSERVER,":{error}"],
	"INFO": [C2S|INTERSERVER,"<server>?"],
	"INVITE": [S2C|C2S,"<nickname> <channel>"],
	"ISON": [C2S,"<nickname>*"],
	"JOIN": [S2C|C2S,"<,|channel>*[ <,|key>*]"],
	"KICK": [S2C|C2S,"<,|channel>* <,|user>*[ :<comment>?]"],
	"KILL": [C2S,"<nickname> :{comment}"],
	"LINKS": [C2S|INTERSERVER,"[[<server>] <mask>]"],
	"LIST": [S2C|C2S,"[<,|channel>*[ <server>]"],
	//Argument order is limit, users, ban masks
	"MODE": [S2C|C2S,"<target> <toggle=+-><modes>[ <args>*]"],
	"NAMES": [C2S,"<,|channel>*"],
	"NICK-1": [S2C|C2S,"<nickname>[ <hopcount>]"],
	"NICK-2": [S2C|C2S,"<nickname>"],
		//^rsp: ["431","432","433","436"],
	"NOTICE": [S2C|C2S,"<target> :{text}?",["text",splitFormattedText]],
	"OPER": [C2S,"<user> <password>"],
	"PART": [C2S|S2C,"<,|channel>*"],
	"PASS": [C2S,"<password>"],
		//^rsp: ["461","462"],
	"PING": [S2C|C2S,"<server>[ <server2>]"],
		//^rsp: ["PONG",/*...*/],
	"PONG": [S2C|C2S,"<daemon>[ <daemon2>]"],
	"PRIVMSG": [S2C|C2S,"<,|target>* :{text}?",["text",splitFormattedText]],
	"QUIT": [S2C|C2S,"[:{message}?]"],
	"REHASH": [C2S,""],
	"RESTART": [C2S,""],
	"SERVER": [INTERSERVER,"<servername> <hopcount> :{info}"],
	"STATS": [C2S|INTERSERVER,"[/query:./[ <server>]]"],
	"SQUIT": [INTERSERVER|C2S,"<server> :{comment}"],
	"SUMMON": [C2S,"<user>[ <server>]"],
	"TIME": [C2S|INTERSERVER,"<server>?"],
	"TOPIC": [C2S|S2C,"<channel>[ :{topic}?]"],
	"TRACE": [C2S|INTERSERVER,"<target>?"],
	"USER-1": [C2S,"<username> <hostname> <servername> <realname>"],
	"USER-2": [C2S,"<username> <mode> * :{realname}"],
		//^rsp: ["461","462"],
	"USERHOST": [C2S,"<nickname>*"],
	"USERS": [C2S|INTERSERVER,"<server>?"],
	"VERSION": [C2S|INTERSERVER,"<server>?"],
	"WALLOPS": [C2S,":{message}"],
	"WHO": [C2S,"[<target>[ /opsOnly:o/]]"],
	"WHOIS": [C2S,"[/server:[^ ]+%.[^ ]+/ ]<,|nickmask>"],
	"WHOWAS": [C2S,"<nickname>[ <#count>[ <server>]]"],
}
//NOTE: Put aliases here
msgs["005-RPL_PROTOCTL"]     = msgs["005-RPL_ISUPPORT"];
msgs["010-RPL_REDIR"]        = msgs["010-RPL_BOUNCE"];
//msgs["999-ERR_NUMERICERR"]   = msgs["999-ERR_NUMERIC_ERR"];

/*
Specs: See http://www.alien.net.au/irc/irc2numerics.html for a summary
When reading a message, it should test all specs unless in strict mode.
When writing a message, it should make note of its spec order in settings
*/
spec = {};
//As described by RFC 1459 (http://www.irchelp.org/irchelp/rfc/rfc.html)
spec.rfc1 = {
	"200": msgs["200-RPL_TRACELINK-1"],
	"201": msgs["201-RPL_TRACECONNECTING"],
	"202": msgs["202-RPL_TRACEHANDSHAKE"],
	"203": msgs["203-RPL_TRACEUNKNOWN"],
	"204": msgs["204-RPL_TRACEOPERATOR"],
	"205": msgs["205-RPL_TRACEUSER"],
	"206": msgs["206-RPL_TRACESERVER-1"],
	"208": msgs["208-RPL_TRACENEWTYPE"],
	//"209": msgs["209-RPL_TRACECLASS"], //reserved
	"211": msgs["211-RPL_STATSLINKINFO"],
	"212": msgs["212-RPL_STATSCOMMANDS"],
	"213": msgs["213-RPL_STATSCLINE"],
	"214": msgs["214-RPL_STATSNLINE"],
	"215": msgs["215-RPL_STATSILINE"],
	"216": msgs["216-RPL_STATSKLINE"],
	//"217": msgs["217-RPL_STATSQLINE"], //reserved
	"218": msgs["218-RPL_STATSYLINE"],
	"219": msgs["219-RPL_ENDOFSTATS"],
	"221": msgs["221-RPL_UMODEIS"],
	//"231": msgs["231-RPL_SERVICEINFO"], //reserved
	//"232": msgs["232-RPL_ENDOFSERVICES"], //reserved
	//"233": msgs["233-RPL_SERVICE"], //reserved
	//"234": msgs["234-RPL_SERVLIST"], //reserved
	//"235": msgs["235-RPL_SERVLISTEND"], //reserved
	"241": msgs["241-RPL_STATSLLINE"],
	"242": msgs["242-RPL_STATSUPTIME"],
	"243": msgs["243-RPL_STATSOLINE"],
	"244": msgs["244-RPL_STATSHLINE"],
	"251": msgs["251-RPL_LUSERCLIENT"],
	"252": msgs["252-RPL_LUSEROP"],
	"253": msgs["253-RPL_LUSERUNKNOWN"],
	"254": msgs["254-RPL_LUSERCHANNELS"],
	"255": msgs["255-RPL_LUSERME"],
	"256": msgs["256-RPL_ADMINME-1"],
	"257": msgs["257-RPL_ADMINLOC1"],
	"258": msgs["258-RPL_ADMINLOC2"],
	"259": msgs["259-RPL_ADMINEMAIL"],
	"261": msgs["261-RPL_TRACELOG"],
	"300": msgs["300-RPL_NONE"],
	"301": msgs["301-RPL_AWAY"],
	"302": msgs["302-RPL_USERHOST"],
	"303": msgs["303-RPL_ISON"],
	"305": msgs["305-RPL_UNAWAY"],
	"306": msgs["306-RPL_NOWAWAY"],
	"311": msgs["311-RPL_WHOISUSER"],
	"312": msgs["312-RPL_WHOISSERVER"],
	"313": msgs["313-RPL_WHOISOPERATOR"],
	"314": msgs["314-RPL_WHOWASUSER"],
	"315": msgs["315-RPL_ENDOFWHO"],
	//"316": msgs["316-RPL_WHOISCHANOP"], //reserved
	"317": msgs["317-RPL_WHOISIDLE"],
	"318": msgs["318-RPL_ENDOFWHOIS"],
	"319": msgs["319-RPL_WHOISCHANNELS"],
	"321": msgs["321-RPL_LISTSTART"],
	"322": msgs["322-RPL_LIST"],
	"323": msgs["323-RPL_LISTEND"],
	"324": msgs["324-RPL_CHANNELMODEIS"],
	"331": msgs["331-RPL_NOTOPIC"],
	"332": msgs["332-RPL_TOPIC"],
	"341": msgs["341-RPL_INVITING-1"],
	"342": msgs["342-RPL_SUMMONING"],
	"351": msgs["351-RPL_VERSION"],
	"352": msgs["352-RPL_WHOREPLY"],
	"353": msgs["353-RPL_NAMREPLY"],
	//"361": msgs["361-RPL_KILLDONE"], //reserved
	//"362": msgs["362-RPL_CLOSING"], //reserved
	//"363": msgs["363-RPL_CLOSEEND"], //reserved
	"364": msgs["364-RPL_LINKS"],
	"365": msgs["365-RPL_ENDOFLINKS"],
	"366": msgs["366-RPL_ENDOFNAMES"],
	"367": msgs["367-RPL_BANLIST"],
	"368": msgs["368-RPL_ENDOFBANLIST"],
	"369": msgs["369-RPL_ENDOFWHOWAS"],
	"371": msgs["371-RPL_INFO"],
	"372": msgs["372-RPL_MOTD"],
	//"373": msgs["373-RPL_INFOSTART"], //reserved
	"374": msgs["374-RPL_ENDOFINFO"],
	"375": msgs["375-RPL_MOTDSTART"],
	"376": msgs["376-RPL_ENDOFMOTD"],
	"381": msgs["381-RPL_YOUREOPER"],
	"382": msgs["382-RPL_REHASHING"],
	//"384": msgs["384-RPL_MYPORTIS"], //reserved
	"391": msgs["391-RPL_TIME-1"],
	"392": msgs["392-RPL_USERSSTART"],
	"393": msgs["393-RPL_USERS"],
	"394": msgs["394-RPL_ENDOFUSERS"],
	"395": msgs["395-RPL_NOUSERS"],
	"401": msgs["401-ERR_NOSUCHNICK"],
	"402": msgs["402-ERR_NOSUCHSERVER"],
	"403": msgs["403-ERR_NOSUCHCHANNEL"],
	"404": msgs["404-ERR_CANNOTSENDTOCHAN"],
	"405": msgs["405-ERR_TOOMANYCHANNELS"],
	"406": msgs["406-ERR_WASNOSUCHNICK"],
	"407": msgs["407-ERR_TOOMANYTARGETS"],
	"409": msgs["409-ERR_NOORIGIN"],
	"411": msgs["411-ERR_NORECIPIENT"],
	"412": msgs["412-ERR_NOTEXTTOSEND"],
	"413": msgs["413-ERR_NOTOPLEVEL"],
	"414": msgs["414-ERR_WILDTOPLEVEL"],
	"421": msgs["421-ERR_UNKNOWNCOMMAND"],
	"422": msgs["422-ERR_NOMOTD"],
	"423": msgs["423-ERR_NOADMININFO"],
	"424": msgs["424-ERR_FILEERROR"],
	"431": msgs["431-ERR_NONICKNAMEGIVEN"],
	"432": msgs["432-ERR_ERRONEUSNICKNAME"],
	"433": msgs["433-ERR_NICKNAMEINUSE"],
	"436": msgs["436-ERR_NICKCOLLISION"],
	"441": msgs["441-ERR_USERNOTINCHANNEL"],
	"442": msgs["442-ERR_NOTONCHANNEL"],
	"443": msgs["443-ERR_USERONCHANNEL"],
	"444": msgs["444-ERR_NOLOGIN"],
	"445": msgs["445-ERR_SUMMONDISABLED"],
	"446": msgs["446-ERR_USERSDISABLED"],
	"451": msgs["451-ERR_NOTREGISTERED"],
	"461": msgs["461-ERR_NEEDMOREPARAMS"],
	"462": msgs["462-ERR_ALREADYREGISTRED"],
	"463": msgs["463-ERR_NOPERMFORHOST"],
	"464": msgs["464-ERR_PASSWDMISMATCH"],
	"465": msgs["465-ERR_YOUREBANNEDCREEP"],
	//"466": msgs["466-ERR_YOUWILLBEBANNED"], //reserved
	"467": msgs["467-ERR_KEYSET"],
	"471": msgs["471-ERR_CHANNELISFULL"],
	"472": msgs["472-ERR_UNKNOWNMODE"],
	"473": msgs["473-ERR_INVITEONLYCHAN"],
	"474": msgs["474-ERR_BANNEDFROMCHAN"],
	"475": msgs["475-ERR_BADCHANNELKEY"],
	//"476": msgs["476-ERR_BADCHANMASK"], //reserved
	"481": msgs["481-ERR_NOPRIVILEGES"],
	"482": msgs["482-ERR_CHANOPRIVSNEEDED"],
	"483": msgs["483-ERR_CANTKILLSERVER"],
	"491": msgs["491-ERR_NOOPERHOST"],
	//"492": msgs["492-ERR_NOSERVICEHOST"], //reserved
	"501": msgs["501-ERR_UMODEUNKNOWNFLAG"],
	"502": msgs["502-ERR_USERSDONTMATCH"],

	"PASS": msgs["PASS"],
	"NICK": msgs["NICK-1"],
	"USER": msgs["USER"],
	"SERVER": msgs["SERVER"],
	"OPER": msgs["OPER"],
	"QUIT": msgs["QUIT"],
	"SQUIT": msgs["SQUIT"],
	"JOIN": msgs["JOIN"],
	"PART": msgs["PART"],
	"MODE": msgs["MODE"],
	"TOPIC": msgs["TOPIC"],
	"NAMES": msgs["NAMES"],
	"LIST": msgs["LIST"],
	"INVITE": msgs["INVITE"],
	"KICK": msgs["KICK"],
	"VERSION": msgs["VERSION"],
	"STATS": msgs["STATS"],
	"LINKS": msgs["LINKS"],
	"TIME": msgs["TIME"],
	"CONNECT": msgs["CONNECT"],
	"TRACE": msgs["TRACE"],
	"ADMIN": msgs["ADMIN"],
	"INFO": msgs["INFO"],
	"PRIVMSG": msgs["PRIVMSG"],
	"NOTICE": msgs["NOTICE"],
	"WHO": msgs["WHO"],
	"WHOIS": msgs["WHOIS"],
	"WHOWAS": msgs["WHOWAS"],
	"KILL": msgs["KILL"],
	"PING": msgs["PING"],
	"PONG": msgs["PONG"],
	"ERROR": msgs["ERROR"],

	"AWAY": msgs["AWAY"],
	"REHASH": msgs["REHASH"],
	"RESTART": msgs["RESTART"],
	"SUMMON": msgs["SUMMON"],
	"USERS": msgs["USERS"],
	"WALLOPS": msgs["WALLOPS"],
	"USERHOST": msgs["USERHOST"],
	"ISON": msgs["ISON"],
};
//spec.rfc1-dep = {}
//As described by RFC 2812 (http://tools.ietf.org/html/rfc2812)
spec.rfc2 = {
	"001": msgs["001-RPL_WELCOME"],
	"NICK": msgs["NICK-2"],
};
//spec.rfc2-dep = {}
//Proposed, unknown origin, etc NOTE: supplemental spec
spec.other = {
	"341": msgs["341-RPL_INVITING-2"],
};
//spec.other-dep = {}
//Extensions by daemons, make note of complete ones (or up to what version they are)
spec.unreal = {};
//spec.unreal-dep = {};
spec.kine = {};
//spec.kine-dep = {};
spec.ircu = {};
//spec.ircu-dep = {};
spec.ircnet = {};
//spec.ircnet-dep = {};
spec.aircd = {};
//spec.aircd-dep = {};
spec.bahamut = {};
//spec.bahamut-dep = {};
spec.hybrid = {};
//spec.hybrid-dep = {};
spec.austhex = {};
//spec.austhex-dep = {};
spec.quakenet = {};
//spec.quakenet-dep = {};
spec.ptlink = {};
//spec.ptlink-dep = {};
spec.ultimate = {};
//spec.ultimate-dep = {};
spec.gamesurge = {};
//spec.gamesurge-dep = {};
spec.bdq = {};
//spec.bdq-dep = {};
spec.undernet = {};
//spec.undernet-dep = {};
spec.ratbox = {};
//spec.ratbox-dep = {};
spec.ithildin = {};
//spec.ithildin-dep = {};

module.exports = {
	msgs: msgs,
	spec: spec,
};

//Make all splitters
for(var x in msgs){
	if(x.charAt(0) == "_"){
		msgs[x] = new Splitter(msgs[x])
	} else {
		msgs[x][1] = new Splitter(msgs[x][1])
		if(msgs[x].length == 2)msgs[x].push(null);
	}
}

for(var spec in spec){
	for(var x in spec[spec]){
		if(typeof(spec[spec][x]) == "undefined"){
			console.log(spec+"."+x+" failed being defined");
		}
	}
}//*/
