// irc/proto.js
/*
Describe messages sent by the server and how to read them
:sender command params\r\n
command is the key, params described in the value (bounds are implied)
Everything's literally matched except:
{name} matches a value potentially with spaces and saves to name (.*)
{name:N} same as above, but may only contain up to N spaces
{name=N} same as above, but must contain exactly N spaces
<name> matches a value without spaces and saves to name ([^ ]*)
<name:etc> same as above, but characters in etc are also not allowed
+Note that to use a literal - instead of a range, use --
<name=etc> match a value that only contains etc. Put a + on the end
+to match multiple of those characters.
<name!etc> same as above but negative.
<#name> matches number and saves to name ([0-9]+)
<#name:.,> matches number with given modifiers ([0-9,]+(\.[0-9]+)?)
+Put an s on the end to switch the meaning of . and ,
<name@otherName> inherets the match rules of otherName. Preceding
+otherName with a _ references the global rule, otherwise references
+one in the current message. Naming is nested for globals.
/name:regex/ specific match (grouping is implied) note: regex is JS's
suffixing with a * (ie. directly after } > or /) will cause the same
+thing to match multiple times and return an array under that name
escape {}<>/%* by preceeding it with a %
Can be an array for multiple forms.
Sections with the same name will assert equivalence within that key
You may denote a section with ()
Sections surrounded by [] denote optional sections.
If there are multiple options per section use [grp1]|[grp2]
If they're not optional, use (grp1)|(grp2)
For all groups - {} <> // () [] - you may start it with one character
+followed by a | to denote what the separator is between multiple
+instances of this section. By default it's space
*/

/*
Message Object {
	sc: S2C | C2S, //sent from Server and/or Client
	//Can be arrays
	msg: "", //Message passed to splitter, default
	rsp: [], //Possible response commands, default
	//The below can also be strings, assumed to be msg
	rfc1459 { //only using rfc1459 (this before default)
		msg: "",
		rsp: [],
	},
	rfc281x: { //with rfc2810-rfc2813 enabled
		msg: "",
		rsp: [],
	},
}
*/

var Splitter = require("splitter.js");

//Server to Client, Client to Server
var S2C=1,C2S=2,INTERSERVER=4,msgs;

function splitWholeEntries(len,data){
	var ret=[];
	while(data.length){
		var rlen = -1;
		for(var i=0;i<data.length && (rlen+=(data[i].length+1)) <= len;++i);
		if(i == 0)
		ret.push(data.slice(0,i));
		data = data.slice(i);
	}
	return ret;
}

function splitFormattedText(len,data){
	var ret = [];
	while(data.length){
		//Make sure to carry over formatting and CTCPs if they can't be completed on that line
		//Word wrap, as well
		//Reference: CTCP \1, Bold \2, Color \3, Italic/Reversed \26, Underline \37, Unformat \17
		var rstr = data.substr(0,len)
		,   rmdr = data.substr(len)
		,   pos,fmt;
		if(rstr.match(/\1/g).length % 2 == 1){
			//CTCP was cut off
			pos = rstr.lastIndexOf("\1");
			rmdr = rstr.substr(pos)+rmdr;
			rstr = rstr.substr(0,pos);
		}
		//Split to last bound (if it's not too far)
		var tmp;
		if(tmp=rstr.match(/(.*[^a-zA-Z0-9\1\2\3\26\37\17].*?)([^\1]*)$/)
		&& tmp[2].length < 20 && !rmdr.match(/^[^a-zA-Z0-9\1\2\3\26\37\17]/)){
			rstr = tmp[1];
			rmdr = tmp[2]+rmdr;
		}
		fmt = rstr.match(/[\2\3\26\37\17]/g);
		if((pos = fmt.lastIndexOf("\17")) != -1){
			//Stopper was found, only retain what comes after
			fmt.slice(pos+1);
		}
		if(fmt.length){
			//Formatting left over, prepend to remainder
			rmdr = fmt.join("")+rmdr;
		}

		ret.push(rstr)
		data = rmdr;
	}
	return ret;
}

/*
Specs: See http://www.alien.net.au/irc/irc2numerics.html for a summary
 * rfc1 - As described by RFC 1459 (http://www.irchelp.org/irchelp/rfc/rfc.html)
 * rfc2 - Added by RFC 2812 (http://tools.ietf.org/html/rfc2812)
 * other - Proposed, unknown, etc
 Extensions by daemons:
 * unreal
 * kine
 * ircu
 * IRCnet
 * aircd
 * bahamut
 * hybrid
 * austhex
 * quakenet
 * ptlink
 * ultimate
 * gamesurge
 * bdq
 * undernet
 * ratbox
 * ithildin
 
When reading a message, it should test all specs unless in strict mode.
When writing a message, it should make note of its spec order in settings

=== Make a note below for which specs are strictly implemented
*/

module.exports = msgs = {
	"_server": ":<sender> <command> {args}",
	"_client": "<command> {args}",
	"_numeric": "<nick> {args}",
	//"_args": "<spaced>* :{trailing}",
	"_target": "(,|"+
		"(/prefix:[#+&]|![A-Z0-9]{5}/<channel:\0\7\r\n,:>[:<mask@channel>])|"+
		"(<user>[%<host>]@<server>)|"+
		"(<user>%<host>)|"+
		"(<nick>[!<user>@<host>])|"+
		//"([$#]"
	")*",
	"_hostname": "<,|@_shortname>*",
	"_shortname": "/:[a-zA-Z0-9][a-zA-Z0-9--][a-zA-Z0-9]/",
	
	//RPL_WELCOME
	"001": {
		sc: S2C,
		rfc2: ":Welcome to the {network} <nick>!<user>@<host>",
	},
	//RPL_YOURHOST
	"002": {
		sc: S2C,
		rfc2: ":Your host is <serverName>, running version <version>",
	},
	//RPL_CREATED
	"003": {
		sc: S2C,
		rfc2: ":This server was created {date}",
	},
	//RPL_MYINFO
	"004": {
		sc: S2C,
		rfc2: "<serverName> <version> <okUserModes> <okChannelModes>",
		kine: "<serverName> <version> <okUserModes> <okChannelModes>"
		+" <paramChannelModes> <paramUserModes> <okServerModes> <paramServerModes>",
	},
	//RPL_ISUPPORT / RPL_PROTOCTL
	"005": [{
		sc: S2C,
		other: "<support>* :are supported by this server", //Sent at the start
		split: ["support",splitWholeEntries],
	},
	//RPL_BOUNCE
	{
		sc: S2C,
		rfc2: ":Try server <server>, port <#port>",
	}],
	
	//RPL_BOUNCE / RPL_REDIR
	"010": {
		sc: S2C,
		other: "<host> <#port> :{info}",
	},
	
	//TODO: All of these with msg only are from rfc1459, check rfc281x for differences.
	//TODO: Remove specific messages, add them to an egnlish language file for them
	
	//RPL_TRACELINK
	"200": { //NOTE: Correct?
		sc: S2C,
		rfc1: "Link <version>[.<debugLevel>] <destination> <nextServer>",
		rfc2: "Link <version>[.<debugLevel>] <destination> <nextServer>"
		+" V<protocolVersion> <linkUptimeSec> <backstreamSendq> <upstreamSendq>",
	},
	//RPL_TRACECONNECTING
	"201": {
		sc: S2C,
		rfc1: "Try. <class> <server>",
	},
	//RPL_TRACEHANDSHAKE
	"202": {
		sc: S2C,
		rfc1: "H.S. <class> <server>",
	},
	//RPL_TRACEUNKNOWN
	"203": {
		sc: S2C,
		rfc1: "???? <class>[ <clientIP>]",
	},
	//RPL_TRACEOPERATOR
	"204": {
		sc: S2C,
		rfc1: "Oper <class> <nick>",
	},
	//RPL_TRACEUSER
	"205": {
		sc: S2C,
		rfc1: "User <class> <nick>",
	},
	//RPL_TRACESERVER
	"206": {
		sc: S2C,
		rfc1: "Serv <class> <#S>S <#C>C <server> <nick>!<user>@<host>",
		rfc2: "Serv <class> <#S>S <#C>C <server> <nick>!<user>@<host> V<protocolVersion>",
	},
	//RPL_TRACESERVICE
	"207": {
		sc: S2C,
		rfc2: "Service <class> <name> <type> <activeType>",
	},
	//RPL_TRACENEWTYPE
	"208": {
		sc: S2C,
		rfc1: "<newtype> 0 <clientName>",
	},
	//RPL_TRACECLASS
	"209": {
		sc: S2C,
		rfc2: "Class <class> <count>",
	},
	//RPL_STATS - ??
	/*"210": {
		sc: S2C,
		aircd: "",
	},*/
	//RPL_STATSLINKINFO
	"211": {
		sc: S2C,
		msg: "<linkname> <sendq> <sentMessages> <sentBytes> <receivedMessages> <receivedBytes> <timeOpen>",
	},
	//RPL_STATSCOMMANDS
	"212": {
		sc: S2C,
		msg: "<command> <count>",
	},
	//RPL_STATSCLINE
	"213": {
		sc: S2C,
		msg: "C <host> * <name> <port> <class>",
	},
	//RPL_STATSNLINE
	"214": {
		sc: S2C,
		msg: "N <host> * <name> <port> <class>",
	},
	//RPL_STATSILINE
	"215": {
		sc: S2C,
		msg: "I <host> * <host> <port> <class>",
	},
	//RPL_STATSKLINE
	"216": {
		sc: S2C,
		msg: "K <host> * <username> <port> <class>",
	},
	//RPL_STATSYLINE
	"218": {
		sc: S2C,
		msg: "Y <class> <pingFrequency> <connectFrequency> <maxSendq>",
	},
	//RPL_ENDOFSTATS
	"219": {
		sc: S2C,
		msg: "<letter> :End of /STATS report",
	},
	//RPL_STATSLLINE
	"241": {
		sc: S2C,
		msg: "L <hostmask> * <servername> <maxdepth>",
	},
	//RPL_STATSUPTIME
	"242": {
		sc: S2C,
		msg: ":Server Up <#days> days <#hour>:<#min>:<#sec>",
	},
	//RPL_STATSOLINE
	"243": {
		sc: S2C,
		msg: "O <hostmask> * <name>",
	},
	//RPL_STATSHLINE
	"244": {
		sc: S2C,
		msg: "H <hostmask> * <servername>",
	},
	//RPL_UMODEIS
	"221": {
		sc: S2C,
		msg: "<userModes>",
	},
	//RPL_LUSERCLIENT
	"251": {
		sc: S2C,
		msg: ":There are <#userCount> users and <#invisCount> invisible on <#servCount> servers",
	},
	//RPL_LUSEROP
	"252": {
		sc: S2C,
		msg: "<#opsCount> :operator%(s%) online",
	},
	//RPL_LUSERUNKNOWN
	"253": {
		sc: S2C,
		msg: "<#unknownConns> :unknown connection%(s%)",
	},
	//RPL_LUSERCHANNELS
	"254": {
		sc: S2C,
		msg: "<#chanCount> :channels formed",
	},
	//RPL_LUSERME
	"255": {
		sc: S2C,
		msg: ":I have <#clients> clients and <#servers> servers",
	},
	//RPL_ADMINME
	"256": [{
		sc: S2C,
		msg: "<server> :Administrative info",
	},{
		sc: S2C,
		msg: "[<server> ]:Administrative info about <server>",
	}]
	//RPL_ADMINLOC1
	"257": {
		sc: S2C,
		msg: ":{adminLocation1}?",
	},
	//RPL_ADMINLOC2
	"258": {
		sc: S2C,
		msg: ":{adminLocation2}?",
	},
	//RPL_ADMINEMAIL
	"259": {
		sc: S2C,
		msg: ":{adminEmail}?",
	},
	//RPL_TRACELOG
	"261": {
		sc: S2C,
		msg: "File <logfile> <debug level>",
	},
	
	//RPL_NONE
	"300": {
		sc: S2C,
		msg: "[{data}]",
	},
	//RPL_AWAY
	"301": {
		sc: S2C,
		msg: "<nick> :{awayMessage}",
	},
	//RPL_USERHOST
	"302": {
		sc: S2C,
		msg: ":[<nick:*=>/op:%*?/=<away=+-><host>]*",
	},
	//RPL_ISON
	"303": {
		sc: S2C,
		msg: ":<nick>*?",
	},
	//RPL_UNAWAY
	"305": {
		sc: S2C,
		msg: ":You are no longer marked as being away",
	},
	//RPL_NOWAWAY
	"306": {
		sc: S2C,
		msg: ":You have been marked as being away",
	},
	//RPL_WHOISUSER
	"311": {
		sc: S2C,
		msg: "<nick> <user> <host> * :{realName}",
	},
	//RPL_WHOISSERVER
	"312": {
		sc: S2C,
		msg: "<nick> <server> :{serverInfo}",
	},
	//RPL_WHOISOPERATOR
	"313": {
		sc: S2C,
		msg: "<nick> :is an IRC operator",
	},
	//RPL_WHOWASUSER
	"314": {
		sc: S2C,
		msg: "<nick> <user> <host> * :{realName}",
	},
	//RPL_ENDOFWHO
	"315": {
		sc: S2C,
		msg: "<name> :End of /WHO list",
	},
	//RPL_WHOISIDLE
	"317": {
		sc: S2C,
		msg: "<nick> <#seconds> :seconds idle",
	},
	//RPL_ENDOFWHOIS
	"318": {
		sc: S2C,
		msg: "<nick> :End of /WHOIS list",
	},
	//RPL_WHOISCHANNELS
	"319": {
		sc: S2C,
		msg: "<nick> :<channel>*",
	},
	//RPL_LISTSTART
	"321": {
		sc: S2C,
		msg: "Channel :Users<= +>Name",
	},
	//RPL_LIST
	"322": {
		sc: S2C,
		msg: "<channel> <#userCount> :{topic}",
	},
	//RPL_LISTEND
	"323": {
		sc: S2C,
		msg: ":End of /LIST",
	},
	//RPL_CHANNELMODEIS
	"324": {
		sc: S2C,
		msg: "<channel> <modes> <modeParams>*",
	},
	//RPL_NOTOPIC
	"331": {
		sc: S2C,
		msg: "<channel> :No topic is set",
	},
	//RPL_TOPIC
	"332": {
		sc: S2C,
		msg: "<channel> :<topic>",
	},
	//RPL_INVITING
	"341": [{
		//NOTE: Use <channel> <nick> in rfc1459 strictness mode, but <nick> <channel> otherwise
		sc: S2C,
		msg: "/channel:[@#+!].*/ <nick>",
	},{
		sc: S2C,
		msg: "<nick> /channel:[@#+!].*/",
	}]
	//RPL_SUMMONING
	"342": {
		sc: S2C,
		msg: "<user> :Summoning user to IRC",
	},
	//RPL_VERSION
	"351": {
		sc: S2C,
		msg: "<version>[.<debuglevel>] <server> :{comments}",
	},
	//RPL_WHOREPLY
	"352": {
		//TODO: What does the asterisk mean? I see an r sometimes, is that the status? What is HG?
		sc: S2C,
		msg: "<channel> <user> <host> <server> <nick> <HG=HG>/asterisk:*//status:.?/ :<hopcount> {realName}",
	},
	//RPL_NAMREPLY
	"353": {
		sc: S2C,
		msg: "<channel> :<nicks>*",
	},
	//RPL_LINKS
	"364": {
		sc: S2C,
		msg: "<mask> <server> :<hopcount> {serverInfo}",
	},
	//RPL_ENDOFLINKS
	"365": {
		sc: S2C,
		msg: "<mask> :End of /LINKS list",
	},
	//RPL_ENDOFNAMES
	"366": {
		sc: S2C,
		msg: "<channel> :End of /NAMES list",
	},
	//RPL_BANLIST
	"367": {
		sc: S2C,
		msg: "<channel> <banid>",
	},
	//RPL_ENDOFBANLIST
	"368": {
		sc: S2C,
		msg: "<channel> :End of channel ban list",
	},
	//RPL_ENDOFWHOWAS
	"369": {
		sc: S2C,
		msg: "<nick> :End of WHOWAS",
	},
	//RPL_INFO
	"371": {
		sc: S2C,
		msg: ":{info}",
	},
	//RPL_MOTD
	"372": {
		sc: S2C,
		msg: ":-[ {motd}]",
	},
	//RPL_ENDOFINFO
	"374": {
		sc: S2C,
		msg: ":End of /INFO list",
	},
	//RPL_MOTDSTART
	"375": {
		sc: S2C,
		msg: ":- <server> Message of the day - ",
	},
	//RPL_ENDOFMOTD
	"376": {
		sc: S2C,
		msg: ":End of /MOTD command",
	},
	//RPL_YOUREOPER
	"381": {
		sc: S2C,
		msg: ":You are now an IRC operator",
	},
	//RPL_REHASHING
	"382": {
		sc: S2C,
		msg: "<configFile> :Rehashing",
	},
	//RPL_TIME
	"391": [{
		sc: S2C,
		msg: "<server> :{timeString}",
	},{ //RPL_TIME: ircu
		sc: S2C,
		msg: "<server> <#timestamp> <#offset> :{timeString}",
	},{ //RPL_TIME: bdq-ircd
		sc: S2C,
		msg: "<server> <timezone> <#microseconds> :{timeString}",
	},{ //RPL_TIME alternative 
		sc: S2C,
		msg: "<server> <#year> <#month> <#day> <#hour> <#minute> #<second>",
	},
	],
	//RPL_USERSSTART
	"392": {
		sc: S2C,
		msg: ":UserID Terminal Host",
	},
	//RPL_USERS
	"393": {
		sc: S2C,
		msg: ":<username> <ttyline> <hostname>",
	},
	//RPL_ENDOFUSERS
	"394": {
		sc: S2C,
		msg: ":End of users",
	},
	//RPL_NOUSERS
	"395": {
		sc: S2C,
		msg: ":Nobody logged in",
	},
	//ERR_NOSUCHNICK
	"401": {
		sc: S2C,
		msg: "<nick> :No such nick/channel",
	},
	//ERR_NOSUCHSERVER
	"402": {
		sc: S2C,
		msg: "<server> :No such server",
	},
	//ERR_NOSUCHCHANNEL
	"403": {
		sc: S2C,
		msg: "<channel> :No such channel",
	},
	//ERR_CANNOTSENDTOCHAN
	"404": {
		sc: S2C,
		msg: "<channel> :Cannot send to channel",
	},
	//ERR_TOOMANYCHANNELS
	"405": {
		sc: S2C,
		msg: "<channel> :You have joined too many channels",
	},
	//ERR_WASNOSUCHNICK
	"406": {
		sc: S2C,
		msg: "<nick> :There was no such nickname",
	},
	//ERR_TOOMANYTARGETS
	"407": {
		sc: S2C,
		msg: "<target> :Duplicate recipients. No message delivered",
	},
	//ERR_NOORIGIN
	"409": {
		sc: S2C,
		msg: ":No origin specified",
	},
	//ERR_NORECIPIENT
	"411": {
		sc: S2C,
		msg: ":No recipient given %(<command>%)",
	},
	//ERR_NOTEXTTOSEND
	"412": {
		sc: S2C,
		msg: ":No text to send",
	},
	//ERR_NOTOPLEVEL
	"413": {
		sc: S2C,
		msg: "<mask> :No toplevel domain specified",
	},
	//ERR_WILDTOPLEVEL
	"414": {
		sc: S2C,
		msg: "<mask> :Wildcard in toplevel domain",
	},
	//ERR_UNKNOWNCOMMAND
	"421": {
		sc: S2C,
		msg: "<command> :Unknown command",
	},
	//ERR_NOMOTD
	"422": {
		sc: S2C,
		msg: ":MOTD File is missing",
	},
	//ERR_NOADMININFO
	"423": {
		sc: S2C,
		msg: "<server> :No administrative info available",
	},
	//ERR_FILEERROR
	"424": {
		sc: S2C,
		msg: ":{reason}",
	},
	//ERR_NONICKNAMEGIVEN
	"431": {
		sc: S2C,
		msg: ":No nickname given",
	},
	//ERR_ERRONEUSNICKNAME
	"432": {
		sc: S2C,
		msg: "<nick> :Erroneus nickname",
	},
	//ERR_NICKNAMEINUSE
	"433": {
		sc: S2C,
		msg: "<nick> :Nickname is already in use",
	},
	//ERR_NICKCOLLISION
	"436": {
		sc: S2C,
		msg: "<nick> :Nickname collision KILL",
	},
	//ERR_USERNOTINCHANNEL
	"441": {
		sc: S2C,
		msg: "<nick> <channel> :They aren't on that channel",
	},
	//ERR_NOTONCHANNEL
	"442": {
		sc: S2C,
		msg: "<channel> :You're not on that channel",
	},
	//ERR_USERONCHANNEL
	"443": {
		sc: S2C,
		msg: "<nick> <channel> :is already on channel",
	},
	//ERR_NOLOGIN
	"444": {
		sc: S2C,
		msg: "<user> :User not logged in",
	},
	//ERR_SUMMONDISABLED
	"445": {
		sc: S2C,
		msg: ":SUMMON has been disabled",
	},
	//ERR_USERSDISABLED
	"446": {
		sc: S2C,
		msg: ":USERS has been disabled",
	},
	//ERR_NOTREGISTERED
	"451": {
		sc: S2C,
		msg: ":You have not registered",
	},
	//ERR_NEEDMOREPARAMS
	"461": {
		sc: S2C,
		msg: "<command> :Not enough parameters",
	},
	//ERR_ALREADYREGISTRED
	"462": {
		sc: S2C,
		msg: ":You may not reregister",
	},
	//ERR_NOPERMFORHOST
	"463": {
		sc: S2C,
		msg: ":Your host isn't among the privileged",
	},
	//ERR_PASSWDMISMATCH
	"464": {
		sc: S2C,
		msg: ":Password incorrect",
	},
	//ERR_YOUREBANNEDCREEP
	"465": {
		sc: S2C,
		msg: ":You are banned from this server",
	},
	//ERR_KEYSET
	"467": {
		sc: S2C,
		msg: "<channel> :Channel key already set",
	},
	//ERR_CHANNELISFULL
	"471": {
		sc: S2C,
		msg: "<channel> :Cannot join channel %(+l%)",
	},
	//ERR_UNKNOWNMODE
	"472": {
		sc: S2C,
		msg: "<char> :is unknown mode char to me",
	},
	//ERR_INVITEONLYCHAN
	"473": {
		sc: S2C,
		msg: "<channel> :Cannot join channel %(+i%)",
	},
	//ERR_BANNEDFROMCHAN
	"474": {
		sc: S2C,
		msg: "<channel> :Cannot join channel %(+b%)",
	},
	//ERR_BADCHANNELKEY
	"475": {
		sc: S2C,
		msg: "<channel> :Cannot join channel %(+k%)",
	},
	//ERR_NOPRIVILEGES
	"481": {
		sc: S2C,
		msg: ":Permission Denied- You're not an IRC operator",
	},
	//ERR_CHANOPRIVSNEEDED
	"482": {
		sc: S2C,
		msg: "<channel> :You're not channel operator",
	},
	//ERR_CANTKILLSERVER
	"483": {
		sc: S2C,
		msg: ":You cant kill a server!",
	},
	//ERR_NOOPERHOST
	"491": {
		sc: S2C,
		msg: ":No O-lines for your host",
	},
	//ERR_UMODEUNKNOWNFLAG
	"501": {
		sc: S2C,
		msg: ":Unknown MODE flag",
	},
	//ERR_USERSDONTMATCH
	"502": {
		sc: S2C,
		msg: ":Cant change mode for other users",
	},
	
	//525-526 - http://bgp.potaroo.net/ietf/all-ids/draft-brocklesby-irc-usercmdpfx-00.txt
	
	//770-774 - http://ithildin.org/trac/browser/trunk/ithildin/modules/ircd/commands/xinfo.c

	"JOIN": {
		sc: S2C|C2S,
		msg: "<,|channel>* <,|key>*",
		rsp: [
			//... a lot
		],
	},
	"KICK": {
		sc: S2C|C2S,
		msg: [
			"<channel> <user>",
			"<channel> <user> :<comment>",
		],
		rsp: [
		],
	},
	"MODE": {
		sc: S2C|C2S,
		//Argument order is limit, users, ban masks
		msg: "<channel> /modes:[+\\-][^ ]+/ <args>*",
		rsp: [
		],
	},
	"NICK": {
		sc: S2C|C2S,
		msg: ["<nickname>","<nickname> <hopcount>"],
		split: ["support",splitWholeEntries],
		rsp: ["431","432","433","436"],
	},
	"NOTICE": {
		sc: S2C|C2S,
		msg: "<target> :{text}",
		split: ["text",splitFormattedText],
	},
	"OPER": {
		sc: C2S,
		msg: "<name> <password>",
	},
	"PART": {
		sc: C2S|S2C,
		msg: "<,|channel>*"
	},
	"PASS": {
		sc: C2S,
		msg: "<password>",
		rsp: ["461","462"],
	},
	"PING": {
		sc: S2C|C2S,
		msg: "<server1> <server2>",
		rsp: ["PONG",//...
		],
	},
	"PONG": {
		sc: S2C|C2S,
		msg: "<daemon1> <daemon2>",
		rsp: [
		],
	},
	"PRIVMSG": {
		sc: S2C|C2S,
		msg: "<target> :{text}",
		split: ["text",splitFormattedText],
	},
	"QUIT": {
		sc: S2C|C2S,
		msg: ":<message>",
	},
	"USER": {
		sc: C2S,
		rfc1459: "<username> <hostname> <servername> <realname>",
		rfc281x: "<username> <mode> * :{realname}",
		rsp: ["461","462"],
	},
}

//Make all splitters
for(var x in msgs){
	var msg = msgs[x];
	if(Array.isArray(msg)){
		for(var i=0;i<msg.length;++i){
			makeSplitter(msg,i,x.charAt(0) == "_");
		}
	} else makeSplitter(msgs,x,x.charAt(0) == "_");
}

function makeSplitter(msgs,x,mode){
	var msg = msgs[x];
	if(mode){
		msgs[x] = new Splitter(msg);
	} else {
		if("msg" in msg){
			msg.msg = new Splitter(msg.msg);
		}
		for(var i=0;i<2;i++){
			var rfc = i==0?"rfc1459":"rfc281x";
			if(rfc in msg){
				if(typeof(msg[rfc]) == "string"){
					msg[rfc] = {
						msg: new Splitter(msg[rfc])
					}
				} else if("msg" in msg[rfc]) {
					msg[rfc].msg = new Splitter(msg[rfc].msg);
				}
			}
			if(!("split" in msg)){
				msg.split = null;
			}
		}
	}
}
